internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -903,31 +969,177 @@ func (r *AgentServiceConfigReconciler) newImageServiceDeployment(ctx context.Con
+				if claim.ObjectMeta.Name == "image-service-data" {
+					found = true
+					statefulSet.Spec.VolumeClaimTemplates[i].Spec.Resources.Requests = getStorageRequests(instance.Spec.ImageStorage)
+				}
+			}
+			if !found {
+				statefulSet.Spec.VolumeClaimTemplates = []corev1.PersistentVolumeClaim{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "image-service-data",
+						},
+						Spec: *instance.Spec.ImageStorage,
+					},
+				}
+			}
+			newVols := make([]corev1.Volume, 0)
+			for i := range volumes {
+				if volumes[i].Name != "image-service-data" {
+					newVols = append(newVols, volumes[i])
+				}
+			}
+			volumes = newVols
+		} else {
+			statefulSet.Spec.VolumeClaimTemplates = nil
+
+			volumes = ensureVolume(volumes, corev1.Volume{
+				Name: "image-service-data",
+				VolumeSource: corev1.VolumeSource{
+					EmptyDir: &corev1.EmptyDirVolumeSource{},
+				},
+			})
+		}
+
+		statefulSet.Spec.Template.Spec.Volumes = volumes
 
 		if r.NodeSelector != nil {
-			deployment.Spec.Template.Spec.NodeSelector = r.NodeSelector
+			statefulSet.Spec.Template.Spec.NodeSelector = r.NodeSelector
 		} else {
-			deployment.Spec.Template.Spec.NodeSelector = map[string]string{}
+			statefulSet.Spec.Template.Spec.NodeSelector = map[string]string{}
 		}
 
 		if r.Tolerations != nil {
-			deployment.Spec.Template.Spec.Tolerations = r.Tolerations
+			statefulSet.Spec.Template.Spec.Tolerations = r.Tolerations
 		} else {
-			deployment.Spec.Template.Spec.Tolerations = []corev1.Toleration{}
+			statefulSet.Spec.Template.Spec.Tolerations = []corev1.Toleration{}
 		}
 		return nil
 	}
 
-	return deployment, mutateFn, nil
+	return statefulSet, mutateFn, nil
+}
+
+func (r *AgentServiceConfigReconciler) reconcileImageServiceStatefulSet(ctx context.Context, log logrus.FieldLogger, instance *aiv1beta1.AgentServiceConfig) error {
+	statefulSet, mutateFn, err := r.newImageServiceStatefulSet(ctx, log, instance)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// delete old deployment if it exists and we've created the new stateful set correctly
+		// TODO: this can be removed when we no longer support upgrading from a release that used deployments for the image service
+		// NOTE: this relies on the err local variable being set correctly when the function returns
+		if err == nil {
+			_ = r.Client.Delete(ctx, &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      imageServiceName,
+					Namespace: r.Namespace,
+				},
+			})
+		}
+	}()
+
+	key := client.ObjectKeyFromObject(statefulSet)
+	if err = r.Client.Get(ctx, key, statefulSet); err != nil {
+		// if the statefulset doesn't exist, create it
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		log.Info("Creating initial image service stateful set")
+		if err = mutateFn(); err != nil {
+			return err
+		}
+		if err = r.Client.Create(ctx, statefulSet); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	existing := statefulSet.DeepCopyObject().(*appsv1.StatefulSet)
```

Comment:
Before doing this, I think we want to check if the `DeletionTimestamp` is set, and if so, return. I assume StatefulSet has a finalizer on it that stays until pods get deleted. So deletion will take some time, and the resource is basically immutable during that time. So there's no point in even attempting an update.